### DISCUSSION TITLE
Why wrap Zustand with React Context?

### DISCUSSION THREAD

**Discussion by KevinMusgrave** [xyflow/xyflow#2667] [Q&A] [answered]
It looks like the zustand state and hooks are not used directly. For example, the main react-flow state is wrapped by a Provider:

https://github.com/wbkd/react-flow/blob/main/packages/core/src/components/ReactFlowProvider/index.tsx

What's the reason for doing this? Doesn't it reduce performance?

**Comment by moklick** [comment, marked as answer]
Hey! 

This is fine. It's also one of the recommended ways to use Zustand https://github.com/pmndrs/zustand#react-context.
We need to do it, because you can have multiple flows with separate stores on one page and we want to give you the possibility to use the Provider higher in your component tree so that you are able to access the store and use the React Flow hooks outside of the main ReactFlow component.

**Comment by RickeyWard** [comment]
Something related to this,

I see the benefit of this is allowing you to have multiple reactflow instances in the same app, but it seems to make it impossible to access the store imperatively from inside your own store.

Actions inside my zustand store need access to width, height, and transform from reactflow useStore. But since its tied to a context, the only way I can see to do this is to duplicate it in my store and listen to the changes and set it on my store but it seems wasteful and convoluted. Though that's how it works for nodes and edges, so I guess that's just the way?
